### PR TITLE
(De)serialize peercrawler

### PR DESCRIPTION
**Pull request by miguel1117**

### CODE REVIEW

File: peercrawler.py [hunk 1]
```diff
@@ -40,15 +41,18 @@ def get_telemetry(ctx, s):
 
 class peer_manager:
     def __init__(self, ctx,
-                 verbosity=0,
+                 verbosity=0, initial_graph: dict[Peer, peer_set] = None,
```
Comment (dsiganos) [inline peercrawler.py:44]: You could set the default to `{}` and avoid the need for an if block.

File: peercrawler.py [hunk 2]
```diff
@@ -325,6 +329,39 @@ def __str__(self):
 
         return s
 
+    def serialize(self) -> str:
```
Comment (dsiganos) [inline peercrawler.py:332]: It would be nice to have the option to serialise to json object.

File: peercrawler.py [hunk 3]
```diff
@@ -325,6 +329,39 @@ def __str__(self):
...
+        for peer, connections in graph_copy.items():
+            peer_data = vars(peer)
+            peer_data["connections"] = [str(id(c)) for c in connections]
+            nodes[id(peer)] = peer_data
+
+        json_connections = json.dumps(nodes, cls=jsonencoder.NanoJSONEncoder)
+        return json_connections
+
+    @staticmethod
+    def deserialize(data: str) -> dict[Peer, peer_set]:
```
Comment (dsiganos) [inline peercrawler.py:344]: It would be nice to be able to deserialise form a json object, not just a json string.

**Review by dsiganos — APPROVED** [review]

File: peercrawler.py (hunk 1 above)
Comment (miguel1117) [reply peercrawler.py:44]: It's really a small detail though, and since I already made it this way it'd be too bothersome to change it now.

**Review by dsiganos — APPROVED** [review]

File: peercrawler.py (hunk 1 above)
Comment (miguel1117) [reply peercrawler.py:44]: Some code is passing `None` to the initial_graph parameter.

**Pull request merged**